### PR TITLE
rgw: check for valid bucket/objects while initializing perms

### PR DESCRIPTION
**Pull request by theanalyst**
Since we use these objects later on, check that they're valid before proceeding

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by dang — APPROVED** [review]

**Comment by cbodley** [comment]
@theanalyst this looks like something we'll want to backport, could you please open a tracker issue?

**Comment by theanalyst** [comment]
> @theanalyst this looks like something we'll want to backport, could you please open a tracker issue?

created https://tracker.ceph.com/issues/57877

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -7958,6 +7958,11 @@ int RGWHandler::do_init_permissions(const DoutPrefixProvider *dpp, optional_yiel
 
 int RGWHandler::do_read_permissions(RGWOp *op, bool only_bucket, optional_yield y)
 {
+  if (op->get_type() != RGW_OP_CREATE_BUCKET &&
+      rgw::sal::Bucket::empty(s->bucket)) {
+    return -ERR_NO_SUCH_BUCKET;
```
Comment (cbodley) [inline src/rgw/rgw_op.cc:7963]: all of the verify jobs in https://pulpito.ceph.com/cbodley-2022-10-18_16:36:54-rgw-wip-cbodley-testing-distro-default-smithi/ failed during run-reshard.sh:

> 2022-10-18T17:57:56.641 INFO:tasks.workunit.client.0.smithi037.stderr:botocore.errorfactory.NoSuchBucket: An error occurred (NoSuchBucket) when calling the ListBuckets operation: Unknown

i think we need to filter out `RGW_OP_LIST_BUCKETS` here too

Comment (theanalyst) [reply src/rgw/rgw_op.cc:7963]: Valid point, I'll rework this!

Comment (theanalyst) [reply src/rgw/rgw_op.cc:7963]: This is helpful, I think I need to filter out all ops that use the service endpoint and don't need a bucket at all. I'll try this and update

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -916,6 +916,16 @@ void rgw_bucket_object_pre_exec(req_state *s)
   dump_bucket_from_state(s);
 }
 
+// TODO: Ideally this should be RGWOp const *, but get_type() isn't
+// const marked
+static bool rgw_is_service_op(RGWOp* op)
```
Comment (cbodley) [inline src/rgw/rgw_op.cc:921]: how about `is_service_op(RGWOpType op)`?

**Comment by cbodley** [comment]
"ceph API tests" is showing some rgw failures that look related

File: src/rgw/rgw_op.cc (hunk 2 above)
Comment (theanalyst) [reply src/rgw/rgw_op.cc:921]: sure much better actually

**Comment by theanalyst** [comment]
> "ceph API tests" is showing some rgw failures that look related

Is there a way to run these tests locally? I tested the few admin apis and swift apis, can check these as well

**Review by adamemerson — APPROVED** [review]

**Comment by theanalyst** [comment]
I changed the approach to only deny on particular object requests for now, since it looks like there are at least a few ops in the admin/roles etc that don't expect a bucket.

**Comment by cbodley** [comment]
can you help me understand how we get into these object ops with empty buckets?

another issue related to these `verify_permission()` functions is that admin/system users ignore these errors and continue on with null buckets (see https://github.com/ceph/ceph/pull/48912)

**Comment by cbodley** [comment]
is it still possible to reproduce this one after the merge of https://github.com/ceph/ceph/pull/49141?

**Comment by theanalyst** [comment]
> is it still possible to reproduce this one after the merge of #49141?

Only the swift path, which is what the new one fixes, everything else is taken care of now.

**Comment by dang** [comment]
Can we close this one?

**Comment by ivancich** [comment]
Teuthology results:
* http://pulpito.ceph.com/ivancich-2023-02-17_21:47:17-rgw-wip-eric-testing-1-distro-default-smithi/
* http://pulpito.ceph.com/ivancich-2023-02-17_21:46:42-rgw-wip-eric-testing-1-distro-default-smithi/

**Comment by ivancich** [comment]
@dang , @cbodley , @theanalyst : I failed to notice discussion was still underway when I ran QA on this. I leave the decision to merge to y'all. I'm adding a DNM until you resolve.

**Comment by ivancich** [comment]
> @dang , @cbodley , @theanalyst : I failed to notice discussion was still underway when I ran QA on this. I leave the decision to merge to y'all. I'm adding a DNM until you resolve.

I should have added you, @adamemerson , to the above. Sorry!

**Comment by theanalyst** [comment]
Closing this one, the Swift POST forms PR is already merged

**Pull request closed without merging**